### ISSUE TITLE
Differences in calling WriteListener.onError

### ISSUE THREAD

**Issue by lachlan-roberts** [jakartaee/servlet#433]
If an async servlet is in a loop like:
```java
    ServletOutputStream out = response.getOutputStream();
    out.setWriteListener(new WriteListener()
    {
        public void onWritePossible()
        {
            try
            {
                while (out.isReady())
                    out.write(someBuffer);
             }
             catch(Throwable t)
             {
                 t.printStackTrace();
             }     
        }
  
        public void onError(Throwable t)
        {
            t.printStackTrace();
        }
    });
```

and the client receiving the response closes the connection at some point, then we see different behaviors from Jetty, Tomcat and Undertow:

- Jetty throws from write and calls onError with a different exception.  If `isReady()` is called it returns true.
- Tomcat throws from write and calls onError with the cause of the original exception.  If `isReady()` is called it returns false.
- Undertow throws from the write, but never calls onError.  If `isReady()` is called it return false, but then `onError` is still not called.

I'm not sure any of these behaviors is strictly correct:
 + should onError be called for an exception that has already been thrown to the application?
 + should onError every be called if `isReady()` has not been called and has not returned false?
 + should the write every actually throw? instead should `isReady()` return false and then `onError` is called with the problem?
 + if the write has thrown, what should `isReady()` return?

**Comment by stuartwdouglas** [comment]
Another alternative would be to delay until IO is attempted, which would match HTTP/1 semantics. That said I don't really like that approach, so I think delivering to the AL is probably the way to go.

**Comment by gregw** [comment]
So if I'm reading you both right.... in async mode:

- mode, write, flush & close may throw an Exception either for the current operation they are called for or for a previous exception if they are called after a previous exception reported either via AL.onError or WL.onError. 
- if the IO operation associated with a write, flush or close (that didn't throw) fails, AND isReady() has not been called, then the exception is reported via AL.onError.
- if the IO operation associated with a write, flush or close (that didn't throw) fails, AND isReady() has been called (and returned false), then the exception is reported via WL.onError and WL.onDataAvailable will never be called. 
- if an IO operation has failed, then isReady() will return false. It will continue to return false even after either onError is called.

I have quibbles with this, but no more or less than other suggestions.  I don't think there is a perfect solution, so happy to go with a common OK solution.

**Comment by stuartwdouglas** [comment]
I was thinking a bit different:


> So if I'm reading you both right.... in async mode:
> 
>     * mode, write, flush & close may throw an Exception either for the current operation they are called for or for a previous exception if they are called after a previous exception reported either via AL.onError or WL.onError.

IMHO write, flush and close should never throw, we should have one place to handle errors and that should be the listener.

> 
>     * if the IO operation associated with a write, flush or close (that didn't throw) fails, AND isReady() has not been called, then the exception is reported via AL.onError.

IMHO these should always be reported to the relevant write listener. In this case because we know the issue is related to a specific write operation then delivering to the write listener error handler makes sense. The isReady stipulation really does not make sense for errors in close().

> 
>     * if the IO operation associated with a write, flush or close (that didn't throw) fails, AND isReady() has been called (and returned false), then the exception is reported via WL.onError and WL.onDataAvailable will never be called.

I think you mean onWritePossible, I agree except for the requirement to call isReady. 

> 
>     * if an IO operation has failed, then isReady() will return false. It will continue to return false even after either onError is called.
> 

Makes sense.

> 
> I have quibbles with this, but no more or less than other suggestions. I don't think there is a perfect solution, so happy to go with a common OK solution.

**Comment by gregw** [comment]
> IMHO write, flush and close should never throw, we should have one place to handle errors and that should be the listener.

yeah but no!!

There are two listeners we can report errors to, so there is no "the listener".      I don't think it ever makes sense to always report to AL.onError (else why have WL.onError), but then it also doesn't work out to always report on WL.onError (but perhaps more sense???).

I too would like write/flush/close to never throw in async.  Perhaps that can be achieved with always calling WL.onError?

Anyway, is my first proposal (3 days ago) better for you?        If not, want to have a go at writing up what you think?

**Comment by stuartwdouglas** [comment]
I think the one from 3 days ago is good. I don't super like the idea of IO exceptions being delivered with no IO operations active but I think it is ok, and is really no different to what happens with timeouts.

**Comment by gregw** [comment]
This is from 3 days ago, but with the addition that write can throw if it is called after a previously reported exception.
 
In async mode:
- mode, write, flush & close should only throw IFF a previous exception has been reported by a call to WL.onError or AL.onError.
- if the IO operation associated with a write, flush or close fails, then it will be reported via the WL.onError callback, regardless of any call/return to isReady
- if an IO operation has failed, then isReady() will return false. It will continue to return false even after onError is called.
- if WL.onError is called, then that will cancel any pending callback of WL.onDataAvailable due to a false return from isReady.
- if there is an IO exception when there is no pending write, flush or close operation, then that is reported via AL.onError

**Comment by stuartwdouglas** [comment]
It looks like there is no circumstances where `RL.onError` can ever be called, due to the last point about it being reported through `AL.onError`. 

I would say that if `isReady` on the ServletInputStream has returned false, the stream is still open (i.e. close has not been called and read() has not returned -1), and there is no pending  write, flush or close operations then we should deliver to RL.onError. 

I also don't like that as written if you attempt to read from the underlying socket, get an IO exception on the read, and there is a pending write operation then you will have to deliver the read exception to the write listener.

**Write:**

- Write, flush & close should only throw IF a previous exception has been reported by a call to WL.onError or AL.onError. (Note that as isReady will always return false write should throw as per the normal isReady semantics).
- If the IO operation associated with a write, flush or close fails, then it will be reported via the WL.onError callback, regardless of any call/return to isReady
- If an IO operation has failed, then isReady() will return false. It will continue to return false even after onError is called.
- If WL.onError is called, then that will cancel any pending callback of WL.onDataAvailable due to a false return from isReady.

**Read:**

- `read` and `close` should never throw, if invoking `read` would cause an exception to be reported then `isReady` should return false, and `onError` invoked.
- A read is considered to be active if `isReady` has returned false. 
- If a read is active then any form of IOException at the transport layer should result in a delivery to `RL.onError`. Note that if a write is also active then this may also be delivered to `WL.onError` as per the rules above. 

**General:**

- If there is an IO exception and according to the rules above it is not delivered to the RL or WL, then it is delivered to the AL. Reasons this could occur include:
- A HTTP/2 RST stream
- Background IO operations detecting a transport layer failure (e.g. the container may be buffering the request in the background, to be ready for a read)
- This potential detection of connection failure is delivered on a best effort basis, applications should not rely on it, as containers will only be able to provide it in specific circumstances.

**Comment by markt-asf** [comment]
I like a lot of this latest description. A couple of minor comments.

In Tomcat, an IOException during `SIS.isReady` will cause `isReady` to return false and `RL.onError` to be called. I think this is consistent with what is written above.

Read needs the same clarification that calling `read` or `close` if `isReady` has returned false will throw as per the standard isReady semantics. Generally, can we make the text for read and write as symmetrical as possible? At the moment it is hard to tell if the differences between them are differences in intended behaviour or just different phrasing.

**Comment by stuartwdouglas** [comment]
I think the Tomcat behavior you describe is consistent with what I have written above, even though it is not explicitly called out. 

I don't think we need to clarify that read will throw after `isReady` returns false, as that is already part of the `isReady` contract. Also I don't think an InputStream `close()` should ever really throw. There has been no historical requirement to have `isReady` return true and IMHO I don't think it makes sense.

In terms of symmetry the text above was a bit of a brain dump. I think we are close to agreement so do you want me to write up a PR tomorrow and we can move the conversation there?

**Comment by markt-asf** [comment]
+1 to a PR,

**Comment by lachlan-roberts** [comment]
I have run the same test with also adding an `AsyncListener` directly before adding the `WriteListener`.

```java
AsyncContext asyncContext = req.startAsync();
asyncContext.addListener(new AsyncListener(){...});
ServletOutputStream out = response.getOutputStream();
out.setWriteListener(new WriteListener()
{
    public void onWritePossible()
    {
        try
        {
            while (out.isReady())
                out.write(someBuffer);
         }
         catch(Throwable t)
         {
             t.printStackTrace();
         }     
    }

    public void onError(Throwable t)
    {
        t.printStackTrace();
        asyncContext.complete();
    }
});
```

- Jetty throws from write and calls onError with a different exception. If isReady() is called it returns true. The `AsyncListener.onComplete()` method is called.

- Tomcat throws from write and calls onError with the cause of the original exception. If isReady() is called it returns false. The `AsyncListener` methods `onError()` and `onComplete()` are both called.

- Undertow throws from the write, but never calls onError. If isReady() is called it return false, but then onError is still not called. No `AsyncListener` methods are ever called.


Note that I have added a `asyncContext.complete()` call to `WriteListener.onError()`. Without this Jetty will never call `AsyncListener.onComplete()`, however it does not change the results of Tomcat and Undertow.

**Comment by stuartwdouglas** [comment]
Undertow will call the `onError` method if `isReady` returns false and then the connection fails while the data is being written out by the IO thread in the background. My thinking was that the exception should only be reported once, but the down side of the current behavior is you have to handle it in two different spots.

The version of Undertow we use in Quarkus (which is basically a fork of Undertow on top of Vert.x) will behave much more like the proposed spec text, with methods never throwing and all errors reported via the listeners.

**Comment by lachlan-roberts** [comment]
I have repeated this experiment after updating my test over Jetty/Tomcat/Undertow to what I believe are the latest servlet 6 versions. (Jetty-12.0.5, Tomcat-10.1.18, Undertow-2.3.10.Final)

The test is setup like this:

The client is continuously reading but aborts when it reads over `1024 * 1024 * 128` bytes.
The server is continuously writing in a loop inside `WriteListener.onWritePossible()`:
```
try
{
    while (outputStream.isReady())
    {
        outputStream.write(data.getBytes());
    }
}
catch (Throwable t)
{
    t.printStackTrace();
}
```



**The results:**

1. Throwing from `ServletOutputStream.write()` inside `onWritePossible()`.

   - Both Jetty and Undertow will always throw from the `ServletOutputStream.write()`.
   - Tomcat will only sometimes throw, if it doesn't throw it will still calls onError after `ServletOutputStream.isReady()` returns false.

2. The result of `ServletOutputStream.isReady()` after throwing from `ServletOutputStream.write()`.

   - Jetty will return true, while Tomcat and Undertow will return false.

3. When is `WriteListener.onError()` and `AsyncListener.onError()` called?

   - Jetty will always only call `WriteListener.onError()`.
   - Tomcat will call both `WriteListener.onError()` then `AsyncListener.onError()`.
   - Undertow will call neither.

4. When is `AsyncListener.onComplete()` called?

   - Always called for Jetty and Tomcat, never called for Undertow.


Note: Even if the exception from `ServletOutputStream.write()` is caught and not re-thrown and `outputStream.isReady()` is not called, then Jetty and Tomcat still invoke the `WriteListener.onError()`.

**Comment by gregw** [comment]
Thanks @lachlan-roberts.
I'll comment a bit here about what think should be the behaviour and then discuss it in more detail in the PR review.

I see no problem with ServletOutputStream.write() throwing even when in async mode.   The reason being that a call to `WriteListener.onError` should only ever be done if there has been a prior call to `isReady()` that returns false.  It is entirely valid for `isReady()` to return true and then for a subsequent call to `write` to throw.  I think all of the following are valid sequences:

1. call to `isReady()` returns true
2. call to `write(...)` throws
3. call to `isReady()` returns true
4. call to `write(...)` throws
5. etc.

OR

1. call to `isReady()` returns true
2. call to `write(...)` returns normally
3. call to `isReady()` returns false
4. `WriteListener.onError` called
5. call to `isReady()` returns true
6. call to `write(...)` throws


I think that `AsyncListener.onError(Throwable)` should only be called if `onWritePossible()` throws, i.e. if it does not catch an exception thrown by `write(...)` or throws some other exception, then that will be passed to `AsyncListener.onError(Throwable)`.  But any exception passed to `WriteListener.onError(Throwable)` or thrown by `write(...)` but caught by the application, should not be passed to `AsyncListener.onError(Throwable)`.




I think the following sequence is not valid:

1. call to `isReady()` returns true
2. call to `write(...)` returns normally
3. call to `isReady()` returns true
4. `WriteListener.onError` called

In short:
 + if `isReady()` returns true, then write may be called, which may throw.
 + if `isReady()` returns false, then write must not be called and either `WL.onWritePossible()` or `WL.onError(Throwable)` will eventually be called.
